In [1]:
import pandas as pd

bp_details = pd.read_csv('bp_details.csv')
all_bp = pd.read_csv('all_bp.csv')
all_oscars = pd.read_csv('all_oscars.csv')
bp_nominees = pd.read_csv('bp_nominees.csv')

In [2]:
#take a look
bp_details.head()
#drop extra index
bp_details= bp_details.drop(columns={'Unnamed: 0'})
bp_details.head()

title  runtime rating                      genre  \
0                    Nomadland  107 min      R                      Drama   
1                   The Father   97 min  PG-13             Drama, Mystery   
2  Judas and the Black Messiah  126 min      R  Biography, Drama, History   
3                         Mank  131 min      R   Biography, Comedy, Drama   
4                       Minari  115 min  PG-13                      Drama   

                                             summary gross  year  
0  A woman in her sixties, after losing everythin...   NaN  2020  
1  A man refuses all assistance from his daughter...   NaN  2020  
2  Offered a plea deal by the FBI, William O'Neal...   NaN  2021  
3  1930s Hollywood is re-evaluated through the ey...   NaN  2020  
4   A Korean family starts a farm in 1980s Arkansas.   NaN  2020

In [3]:
#all_bp = all_bp.drop(columns={'Unnamed: 0'})
all_bp = all_bp.query('year >= 1963')
all_bp = all_bp.query('year <= 2020')
all_bp = all_bp.drop(columns = {'Unnamed: 0'})
all_bp

category                           nominee  year  winner
0    Best Picture                Lawrence of Arabia  1963     1.0
1    Best Picture                   The Longest Day  1963     0.0
2    Best Picture  Meredith Willson's The Music Man  1963     0.0
3    Best Picture              Mutiny on the Bounty  1963     0.0
4    Best Picture             To Kill a Mockingbird  1963     0.0
..            ...                               ...   ...     ...
328  Best Picture                             Joker  2020     0.0
329  Best Picture                      Little Women  2020     0.0
330  Best Picture                    Marriage Story  2020     0.0
331  Best Picture                              1917  2020     0.0
332  Best Picture   Once upon a Time...in Hollywood  2020     0.0

[333 rows x 4 columns]

In [4]:
all_oscars = all_oscars.drop(columns={'Unnamed: 0'})
all_oscars.head()

category              nominee  year  winner
0    Actor        Emil Jannings  1929     1.0
1    Actor  Richard Barthelmess  1929     0.0
2  Actress         Janet Gaynor  1929     1.0
3  Actress       Louise Dresser  1929     0.0
4  Actress       Gloria Swanson  1929     0.0

In [5]:
bp_nominees = bp_nominees.drop(columns={'Unnamed: 0'})
bp_nominees

film           film_studio film_studio/producer(s)  \
0                     NaN                   NaN                     NaN   
1                   Wings  Famous Players-Lasky                     NaN   
2              7th Heaven                   Fox                     NaN   
3              The Racket     The Caddo Company                     NaN   
4                     NaN                   NaN                     NaN   
..                    ...                   ...                     ...   
670          King Richard                   NaN                     NaN   
671        Licorice Pizza                   NaN                     NaN   
672       Nightmare Alley                   NaN                     NaN   
673  The Power of the Dog                   NaN                     NaN   
674       West Side Story                   NaN                     NaN   

                                           producer(s) films  oscars_year  
0                                                  NaN   NaN         1927  
1                                                  NaN   NaN         1927  
2                                                  NaN   NaN         1927  
3                                                  NaN   NaN         1927  
4                                                  NaN   NaN         1928  
..                                                 ...   ...          ...  
670            Tim White, Trevor White, and Will Smith   NaN         2021  
671  Sara Murphy, Adam Somner, and Paul Thomas Ande...   NaN         2021  
672  J. Miles Dale, Guillermo del Toro, and Bradley...   NaN         2021  
673  Emile Sherman, Iain Canning, Roger Frappier, J...   NaN         2021  
674       Steven Spielberg and Kristie Macosko Krieger   NaN         2021  

[675 rows x 6 columns]

In [6]:
#box office numbers and gross aren't available yet for 2020, Best Picture became a category in 1963, so narrow in on years 1963-2020

bp_details = bp_details.query('year >= 1962')
bp_details = bp_details.query('year <= 2019')
bp_details

title  runtime     rating                        genre  \
6           Sound of Metal  120 min          R                 Drama, Music   
8                 Parasite  132 min          R      Comedy, Drama, Thriller   
9           Ford v Ferrari  152 min      PG-13     Action, Biography, Drama   
10                   Joker  122 min          R       Crime, Drama, Thriller   
11          Marriage Story  137 min          R               Drama, Romance   
..                     ...      ...        ...                          ...   
335     Lawrence of Arabia  218 min   Approved  Adventure, Biography, Drama   
336        The Longest Day  178 min          G       Action, Drama, History   
337          The Music Man  151 min          G      Comedy, Family, Musical   
338   Mutiny on the Bounty  178 min  Not Rated    Adventure, Drama, History   
339  To Kill a Mockingbird  129 min   Approved                 Crime, Drama   

                                               summary        gross  year  
6    A heavy-metal drummer's life is thrown into fr...          NaN  2019  
8    Greed and class discrimination threaten the ne...   53,367,844  2019  
9    American car designer Carroll Shelby and drive...  117,624,028  2019  
10   A mentally troubled stand-up comedian embarks ...  335,451,311  2019  
11   Noah Baumbach's incisive and compassionate loo...    2,000,000  2019  
..                                                 ...          ...   ...  
335  The story of T.E. Lawrence, the English office...   44,824,144  1962  
336  The events of D-Day, told on a grand scale fro...   39,100,000  1962  
337  Traveling con artist Harold Hill targets the n...   14,953,846  1962  
338  In 1787, British ship Bounty leaves Portsmouth...   13,680,000  1962  
339  Atticus Finch, a widowed lawyer in Depression-...          NaN  1962  

[332 rows x 7 columns]

In [7]:
all_bp = all_bp.query('year >= 1963')
all_bp = all_bp.query('year <= 2021')
all_bp = all_bp.reset_index(drop=True)
all_bp

category                           nominee  year  winner
0    Best Picture                Lawrence of Arabia  1963     1.0
1    Best Picture                   The Longest Day  1963     0.0
2    Best Picture  Meredith Willson's The Music Man  1963     0.0
3    Best Picture              Mutiny on the Bounty  1963     0.0
4    Best Picture             To Kill a Mockingbird  1963     0.0
..            ...                               ...   ...     ...
328  Best Picture                             Joker  2020     0.0
329  Best Picture                      Little Women  2020     0.0
330  Best Picture                    Marriage Story  2020     0.0
331  Best Picture                              1917  2020     0.0
332  Best Picture   Once upon a Time...in Hollywood  2020     0.0

[333 rows x 4 columns]

test = pd.merge(bp_details, all_bp, left_on='title', right_on='nominee', how='outer')
test.head(60)

test[test.title.isna()]

#test merge shows some titles are different in each database - 368 in outer merge vs 303 in inner merge - 338 in bp_details vs 333 in bp_nominees
#Music Man vs Meredith Wilson's Music Man
#Who's Afraid of Virginia Woolf (?)
#MASH no special *
#Cries and (&) Whispers
#1917 gone from bp_details (reg ex error)
#Precious
#birdman
#127 hours (regex error)
#Once (...) upon a time in Hollywood

bp_details = bp_details.sort_values(['year', 'title']).reset_index(drop=True)
bp_details

all_bp = all_bp.sort_values(['year', 'nominee']).reset_index(drop=True)
all_bp

pd.concat([bp_details, all_bp], axis=1)

test2 = pd.merge(bp_details, all_bp, left_on='title', right_on='nominee', how='inner')
test2 = test2.rename(columns={'year_x':'premiere_year', 'year_y':'oscar_year'})
test2

!conda install -c conda-forge fuzzywuzzy

In [8]:
import difflib
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

def fuzzy_merge(df_1, df_2, key1, key2, threshold=90, limit=2):

#df_1 is the left table to join
#df_2 is the right table to join
#key1 is the key column of the left table
#key2 is the key column of the right table
#threshold is how close the matches should be to return a match, based on Levenshtein distance
#limit is the amount of matches that will get returned, these are sorted high to low

    s = df_2[key2].tolist()

    m = df_1[key1].apply(lambda x: process.extract(x, s, limit=limit))    
    df_1['matches'] = m

    m2 = df_1['matches'].apply(lambda x: ', '.join([i[0] for i in x if i[1] >= threshold]))
    df_1['matches'] = m2

    return df_1

In [9]:
big_df = fuzzy_merge(bp_details, all_bp, 'title', 'nominee', threshold=90, limit=1)
big_df

title  runtime     rating                        genre  \
6           Sound of Metal  120 min          R                 Drama, Music   
8                 Parasite  132 min          R      Comedy, Drama, Thriller   
9           Ford v Ferrari  152 min      PG-13     Action, Biography, Drama   
10                   Joker  122 min          R       Crime, Drama, Thriller   
11          Marriage Story  137 min          R               Drama, Romance   
..                     ...      ...        ...                          ...   
335     Lawrence of Arabia  218 min   Approved  Adventure, Biography, Drama   
336        The Longest Day  178 min          G       Action, Drama, History   
337          The Music Man  151 min          G      Comedy, Family, Musical   
338   Mutiny on the Bounty  178 min  Not Rated    Adventure, Drama, History   
339  To Kill a Mockingbird  129 min   Approved                 Crime, Drama   

                                               summary        gross  year  \
6    A heavy-metal drummer's life is thrown into fr...          NaN  2019   
8    Greed and class discrimination threaten the ne...   53,367,844  2019   
9    American car designer Carroll Shelby and drive...  117,624,028  2019   
10   A mentally troubled stand-up comedian embarks ...  335,451,311  2019   
11   Noah Baumbach's incisive and compassionate loo...    2,000,000  2019   
..                                                 ...          ...   ...   
335  The story of T.E. Lawrence, the English office...   44,824,144  1962   
336  The events of D-Day, told on a grand scale fro...   39,100,000  1962   
337  Traveling con artist Harold Hill targets the n...   14,953,846  1962   
338  In 1787, British ship Bounty leaves Portsmouth...   13,680,000  1962   
339  Atticus Finch, a widowed lawyer in Depression-...          NaN  1962   

                              matches  
6                                      
8                            Parasite  
9                      Ford v Ferrari  
10                              Joker  
11                     Marriage Story  
..                                ...  
335                Lawrence of Arabia  
336                   The Longest Day  
337  Meredith Willson's The Music Man  
338              Mutiny on the Bounty  
339             To Kill a Mockingbird  

[332 rows x 8 columns]

In [10]:
everything = pd.merge(big_df, all_bp, left_on='matches', right_on='nominee', how='outer')

In [11]:
everything.tail(20)

title  runtime     rating  \
317                                       My Fair Lady  170 min          G   
318                                             Becket  148 min      PG-13   
319  Dr. Strangelove or: How I Learned to Stop Worr...   95 min         PG   
320                                       Mary Poppins  139 min          G   
321                                    Zorba the Greek  142 min  Not Rated   
322                                          Tom Jones  129 min  Not Rated   
323                                    America America  174 min   Approved   
324                                          Cleopatra  192 min          G   
325                               How the West Was Won  164 min          G   
326                                Lilies of the Field   94 min    Unrated   
327                                 Lawrence of Arabia  218 min   Approved   
328                                    The Longest Day  178 min          G   
329                                      The Music Man  151 min          G   
330                               Mutiny on the Bounty  178 min  Not Rated   
331                              To Kill a Mockingbird  129 min   Approved   
332                                                NaN      NaN        NaN   
333                                                NaN      NaN        NaN   
334                                                NaN      NaN        NaN   
335                                                NaN      NaN        NaN   
336                                                NaN      NaN        NaN   

                           genre  \
317       Drama, Family, Musical   
318    Biography, Drama, History   
319                  Comedy, War   
320      Comedy, Family, Fantasy   
321                Comedy, Drama   
322   Adventure, Comedy, History   
323                        Drama   
324    Biography, Drama, History   
325                      Western   
326                        Drama   
327  Adventure, Biography, Drama   
328       Action, Drama, History   
329      Comedy, Family, Musical   
330    Adventure, Drama, History   
331                 Crime, Drama   
332                          NaN   
333                          NaN   
334                          NaN   
335                          NaN   
336                          NaN   

                                               summary        gross  year_x  \
317  In 1910s London, snobbish phonetics professor ...   72,000,000  1964.0   
318  King Henry II of England comes to terms with h...          NaN  1964.0   
319  An insane American general orders a bombing at...      275,902  1964.0   
320  In turn of the century London, a magical nanny...  102,272,727  1964.0   
321  An uptight English writer travelling to Crete,...          NaN  1964.0   
322  The romantic and chivalrous adventures of adop...   37,600,000  1963.0   
323  A young Greek stops at nothing to secure a pas...          NaN  1963.0   
324  Queen Cleopatra VII of Egypt experiences both ...   57,777,778  1963.0   
325  A family saga covering several decades of West...   46,500,000  1962.0   
326  A travelling handyman becomes the answer to th...          NaN  1963.0   
327  The story of T.E. Lawrence, the English office...   44,824,144  1962.0   
328  The events of D-Day, told on a grand scale fro...   39,100,000  1962.0   
329  Traveling con artist Harold Hill targets the n...   14,953,846  1962.0   
330  In 1787, British ship Bounty leaves Portsmouth...   13,680,000  1962.0   
331  Atticus Finch, a widowed lawyer in Depression-...          NaN  1962.0   
332                                                NaN          NaN     NaN   
333                                                NaN          NaN     NaN   
334                                                NaN          NaN     NaN   
335                                                NaN          NaN     NaN   
336                                                NaN          NaN     N

In [12]:
everything.to_csv('bp_full_df1.csv')